zerver/actions/message_delete.py

```diff
@@ -15,6 +15,32 @@ class DeleteMessagesEvent(TypedDict, total=False):
     stream_id: int
 
 
+def check_update_first_message_id(
+    realm: Realm, stream: Stream, message_ids: List[int], users_to_notify: Iterable[int]
+) -> None:
+    # This will not update the `first_message_id` of streams where the
+    # first message was deleted prior to the implementation of this function.
+    if not stream.recipient_id or stream.first_message_id not in message_ids:
+        return
+    stream_messages = Message.objects.filter(
+        realm_id=realm.id, recipient_id=stream.recipient_id
+    ).order_by("date_sent")
```

Comment:
Order by ID -- that's the primary ordering key for messages; `date_sent` order should be basically always the same, but it's much less nice for indexes than the integer `ID` key.

Reply:
Also, use `.first()` to ask the database to to return the first result, if any, or `None` otherwise, without multiple queries, which is what your code will do.

Worth your reading the Django QuerySet documentation a bit.

Reply:
OK, I understand now - using QuerySet in a boolean context triggers query fetch, right? I have fixed it now.

Reply:
Yeah, exactly, it fetches the full set of results.